lib/node/runner/checker.go

```diff
@@ -556,64 +560,116 @@ func TransitStateToACCEPT(c common.Checker, args ...interface{}) (err error) {
 
 // FinishedBallotStore will store the confirmed ballot to
 // `Block`.
-func FinishedBallotStore(c common.Checker, args ...interface{}) (err error) {
+func FinishedBallotStore(c common.Checker, args ...interface{}) error {
 	checker := c.(*BallotChecker)
 
 	if !checker.VotingFinished {
-		return
+		return nil
 	}
-	ballotRound := checker.Ballot.VotingBasis()
-	if checker.FinishedVotingHole == voting.YES {
-		if err = getMissingTransaction(checker); err != nil {
-			checker.Log.Debug("failed to get the missing transactions of ballot", "error", err)
-			return
-		}
 
-		var theBlock *block.Block
+	basis := checker.Ballot.VotingBasis()
 
-		var bs *storage.LevelDBBackend
-		if bs, err = checker.NodeRunner.Storage().OpenBatch(); err != nil {
+	var err error
+	switch checker.FinishedVotingHole {
+	case voting.YES:
+		if err = saveBlock(checker); err != nil {
 			return err
 		}
+		checker.NodeRunner.TransitISAACState(checker.Ballot.VotingBasis(), ballot.StateALLCONFIRM)
+		checker.NodeRunner.Consensus().SetLatestVotingBasis(basis)
 
-		theBlock, err = finishBallot(
-			bs,
-			checker.Ballot,
-			checker.NodeRunner.TransactionPool,
-			checker.Log,
-			checker.NodeRunner.Log(),
-		)
-		if err != nil {
-			bs.Discard()
-			checker.Log.Error("failed to finish ballot", "error", err)
-			return
-		}
-
-		if err = bs.Commit(); err != nil {
-			if err != errors.NotCommittable {
-				bs.Discard()
-				return
-			}
-		}
-
-		checker.Log.Debug("ballot was stored", "block", *theBlock)
-		checker.NodeRunner.SavingBlockOperations().Save(*theBlock)
-		checker.NodeRunner.TransitISAACState(ballotRound, ballot.StateALLCONFIRM)
+		reorganizeTransactionPool(checker)
+		checker.NodeRunner.Consensus().RemoveRunningRoundsWithSameHeight(basis.Height)
 
 		err = NewCheckerStopCloseConsensus(checker, "ballot got consensus and will be stored")
-	} else {
+	case voting.NO, voting.EXP:
 		checker.NodeRunner.isaacStateManager.IncreaseRound()
+		checker.NodeRunner.Consensus().SetLatestVotingBasis(basis)
+
+		checker.NodeRunner.Consensus().RemoveRunningRoundsWithSameHeight(basis.Height)
+
 		err = NewCheckerStopCloseConsensus(checker, "ballot got consensus")
+	case voting.NOTYET:
+		return errors.New("invalid voting.Hole, `NOTYET`")
 	}
 
-	checker.NodeRunner.Consensus().CloseConsensus(
-		checker.Ballot.Proposer(),
-		ballotRound,
-		checker.FinishedVotingHole,
+	return err
+}
+
+func saveBlock(checker *BallotChecker) error {
+	var err error
+	if err = insertMissingTransaction(checker); err != nil {
+		checker.Log.Debug("failed to get the missing transactions of ballot", "error", err)
+		return err
+	}
+
+	var bs *storage.LevelDBBackend
+	if bs, err = checker.NodeRunner.Storage().OpenBatch(); err != nil {
+		return err
+	}
+
+	proposedTransactions, err := getProposedTransactions(
+		bs,
+		checker.Ballot.B.Proposed.Transactions,
 		checker.NodeRunner.TransactionPool,
 	)
+	if err != nil {
+		return err
+	}
 
-	return
+	for _, tx := range proposedTransactions {
```

Comment:
So, this piece of code is almost like `finishBallot` but it uses `proposedTransactions` to mark `LatestUpdatedSources`. Unfortunately this is done before the block is created, which means we could mark them as updated yet have an error. Is it intentional ?
Also, it makes updating the code quite harder, since we now have 2 pieces of code which do the same thing but not quite.